### PR TITLE
Moved logic for pod jobs from Sinker to Plank

### PR DESCRIPTION
**Pull request by hpandeycodeit**
Moved logic for pod from Sinker to Plank as per Issue #13706

Old PR: #13863

### CODE REVIEW

File: prow/plank/controller.go [hunk 1]
```diff
@@ -28,10 +28,11 @@ import (
 	"github.com/sirupsen/logrus"
 	coreapi "k8s.io/api/core/v1"
 	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
```
Comment (alvaroaleman) [inline prow/plank/controller.go]: The same import already exists one line below under a different alias :)

File: prow/plank/controller.go [hunk 2]
```diff
@@ -98,6 +103,16 @@ type Controller struct {
 
 	// if skip report job results to github
 	skipReport bool
+
+	//prowJob Client
+	//prowJobClient prowv1.ProwJobInterface
+
+	// pod client
+	podClients []corev1.PodInterface
```
Comment (alvaroaleman) [inline prow/plank/controller.go]: These clients are already available as `buildClients` on line 83

File: prow/plank/controller.go [hunk 3]
```diff
@@ -591,4 +607,123 @@ func (c *Controller) patchProwjob(srcPJ prowapi.ProwJob, destPJ prowapi.ProwJob)
 	_, err = c.prowJobClient.Patch(srcPJ.Name, ktypes.MergePatchType, patch)
 	c.log.WithFields(pjutil.ProwJobFields(&destPJ)).Debug("Patched ProwJob.")
 	return err
+
+}
+
+type plankReconciliationMetrics struct {
```
Comment (alvaroaleman) [inline prow/plank/controller.go]: Can we remove the metrics we move over to plank from sinker?

Also the `prowJobsXXX` metrics don't make sense here, as the ProwJob cleanup still remains in sinker

File: prow/plank/controller.go [hunk 4]
```diff
@@ -591,4 +607,123 @@ func (c *Controller) patchProwjob(srcPJ prowapi.ProwJob, destPJ prowapi.ProwJob)
...
+	podRemovalErrors       map[string]int
+	prowJobsCreated        int
+	prowJobsCleaned        map[string]int
+	prowJobsCleaningErrors map[string]int
+}
+
+// Get the jobs that we need to retain so horologium can continue working
+// as intended.
+// And clean up old pods.
+func (c *Controller) managePodJobs(pj prowapi.ProwJob) {
```
Comment (alvaroaleman) [inline prow/plank/controller.go]: Can we rename this to something more intuitive, e.G. `cleanUpPods`?

File: prow/plank/controller.go [hunk 5]
```diff
@@ -530,6 +545,7 @@ func (c *Controller) syncTriggeredJob(pj prowapi.ProwJob, pm map[string]coreapi.
 			WithField("from", prevState).
 			WithField("to", pj.Status.State).Info("Transitioning states.")
 	}
+	c.managePodJobs(pj)
```
Comment (alvaroaleman) [inline prow/plank/controller.go]: This doesn't really make sense here, we now run a pod cleanup once for each prowjob that is triggered, but we want to run it only once.  Putting it on line 278 would probably make sense

File: prow/plank/controller.go [hunk 6]
```diff
@@ -28,10 +28,11 @@ import (
 	"github.com/sirupsen/logrus"
```
Comment (alvaroaleman) [inline prow/plank/controller.go]: Please also move the tests over

**Comment by matthyx** [comment]
Please run `hack/update-deps.sh` or even `hack/update-all.sh` before updating your PR.

File: prow/plank/controller_test.go [hunk 7]
```diff
@@ -1684,3 +1685,203 @@ func TestMaxConcurency(t *testing.T) {
...
+	deletedPodsTrusted := sets.NewString("old-failed-trusted")
+
+	fkc := []*fake.Clientset{fake.NewSimpleClientset(pods...), fake.NewSimpleClientset(podsTrusted...)}
+	var fpc []corev1.PodInterface
+	for _, fakeClient := range fkc {
+		fpc = append(fpc, fakeClient.CoreV1().Pods("ns"))
+	}
+	c := Controller{
+		podClients: fpc,
+		config:     newFakeConfigAgent(pods, 0).Config,
```
Comment (matthyx) [inline prow/plank/controller_test.go]: should be something like:
`config:     newFakeConfigAgent(t, 0).Config,`

File: prow/plank/controller.go [hunk 8]
```diff
@@ -598,4 +607,49 @@ func (c *Controller) patchProwjob(srcPJ prowapi.ProwJob, destPJ prowapi.ProwJob)
...
+
+	// Now clean up old pods.
+	selector := fmt.Sprintf("%s = %s", kube.CreatedByProw, "true")
+	for _, client := range c.podClients {
+		pods, err := client.List(metav1.ListOptions{LabelSelector: selector})
+		if err != nil {
+			c.log.WithError(err).Error("Error listing pods.")
+			return
+		}
+		maxPodAge := c.config().Sinker.MaxPodAge.Duration
```
Comment (matthyx) [inline prow/plank/controller.go]: Your test is failing here, because the `newFakeConfigAgent` is missing Sinker config...

Comment (matthyx) [reply prow/plank/controller.go]: You should fix it in `controller_test.go` at the beginning of the file.

**Comment by matthyx** [comment]
Please run `hack/update-gofmt.sh`

**Comment by matthyx** [comment]
Thanks, now it should be ready for review!
/assign @cblecker

**Comment by hpandeycodeit** [comment]
@matthyx Thank you for the help

File: prow/plank/controller.go (hunk 2 above)
Comment (alvaroaleman) [reply prow/plank/controller.go]: This comment still applies, please remove the duplicate clients here and use the `.buildClients` property from the `Controller`

File: prow/cmd/sinker/main.go [hunk 9]
```diff
@@ -360,45 +357,6 @@ func (c *controller) clean() {
 		}
 	}
 
-	// Now clean up old pods.
```
Comment (alvaroaleman) [inline prow/cmd/sinker/main.go]: I think we can completely get rid of the pod clients in sinker now (l133-142 & remove the `podClient` property from the `controller` struct)

File: prow/plank/controller.go [hunk 10]
```diff
@@ -598,4 +606,49 @@ func (c *Controller) patchProwjob(srcPJ prowapi.ProwJob, destPJ prowapi.ProwJob)
 	_, err = c.prowJobClient.Patch(srcPJ.Name, ktypes.MergePatchType, patch)
 	c.log.WithFields(pjutil.ProwJobFields(&destPJ)).Debug("Patched ProwJob.")
 	return err
+
+}
+
+// Get the jobs that we need to retain so horologium can continue working
+// as intended.
+// And clean up old pods.
+func (c *Controller) cleanUpPods() {
```
Comment (alvaroaleman) [inline prow/plank/controller.go]: Please also move the metrics (podsCreated, podsRemoved, podsRemovalErrors) over instead of just removing them

**Comment by matthyx** [comment]
Please fix:
`/b/f/w/prow/cmd/sinker/main.go:30:2: imported and not used: "k8s.io/client-go/kubernetes/typed/core/v1" as corev1`

**Pull request closed without merging**